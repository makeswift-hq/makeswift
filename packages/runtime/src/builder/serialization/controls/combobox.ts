import { ComboboxControlDefinition } from '../../../controls'
import { deserializeFunction, serializeFunction } from '../function-serialization'
import { Deserialize, Serialize } from './types'

export function serializeComboboxControlDefinition(
  definition: ComboboxControlDefinition,
): [Serialize<ComboboxControlDefinition>, Transferable[]] {
  const transferables = []
  const getOptions = definition.config.getOptions && serializeFunction(definition.config.getOptions)

  if (getOptions) transferables.push(getOptions)

  return [{ ...definition, config: { ...definition.config, getOptions } }, transferables] as [
    Serialize<ComboboxControlDefinition>,
    Transferable[],
  ]
}

export function deserializeComboboxControlDefinition(
  definition: Serialize<ComboboxControlDefinition>,
): Deserialize<Serialize<ComboboxControlDefinition>> {
  const getOptions =
    definition.config.getOptions && deserializeFunction(definition.config.getOptions)

  return {
    ...definition,
    config: { ...definition.config, getOptions },
  } as Deserialize<Serialize<ComboboxControlDefinition>>
}
