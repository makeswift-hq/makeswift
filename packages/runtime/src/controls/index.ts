export * from './color'
export * from './combobox'
export * from './control'
export * from './icon-radio-group'
export * from './image'
export * from './link'
export * from './list'
export * from './select'
export * from './shape'
export * from './slot'
export * from './style'
export * from './style-v2'
export * from './rich-text'
export * from './rich-text-v2'
export * from './typography'
