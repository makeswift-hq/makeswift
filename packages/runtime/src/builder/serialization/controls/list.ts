import { ListControlDefinition } from '../../../controls'
import { deserializeControl, serializeControl } from '../control-serialization'
import { deserializeFunction, serializeFunction } from '../function-serialization'
import { Deserialize, Serialize } from './types'

export function serializeListControlDefinition(
  definition: ListControlDefinition,
): [Serialize<ListControlDefinition>, Transferable[]] {
  const [type, transferables] = serializeControl(definition.config.type)
  const getItemLabel =
    definition.config.getItemLabel && serializeFunction(definition.config.getItemLabel)

  if (getItemLabel) transferables.push(getItemLabel)

  return [
    { ...definition, config: { ...definition.config, type, getItemLabel } },
    transferables,
  ] as [Serialize<ListControlDefinition>, Transferable[]]
}

export function deserializeListControlDefinition(
  definition: Serialize<ListControlDefinition>,
): Deserialize<Serialize<ListControlDefinition>> {
  const type = deserializeControl(definition.config.type)
  const getItemLabel =
    definition.config.getItemLabel && deserializeFunction(definition.config.getItemLabel)

  return {
    ...definition,
    config: { ...definition.config, type, getItemLabel },
  } as Deserialize<Serialize<ListControlDefinition>>
}
