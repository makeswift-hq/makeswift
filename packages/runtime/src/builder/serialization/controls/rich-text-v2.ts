import { RichTextV2ControlDefinition } from '../../../controls'
import { Deserialize, Serialize } from './types'

export function serializeRichTextControlV2(
  control: RichTextV2ControlDefinition,
): [Serialize<RichTextV2ControlDefinition>, Transferable[]] {
  const { plugins, ...config } = control.config
  const transferables: Transferable[] = []

  return [
    {
      ...control,
      config: {
        ...config,
        plugins:
          plugins?.map(plugin => {
            return {
              control: {
                definition: plugin?.control?.definition,
              },
            }
          }) ?? [],
      },
    },
    transferables,
  ] as [Serialize<RichTextV2ControlDefinition>, Transferable[]]
}

export function deserializeRichTextControlV2(
  control: Serialize<RichTextV2ControlDefinition>,
): Deserialize<Serialize<RichTextV2ControlDefinition>> {
  return control as Deserialize<Serialize<RichTextV2ControlDefinition>>
}
