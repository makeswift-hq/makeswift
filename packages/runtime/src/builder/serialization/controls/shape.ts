import { ShapeControlDefinition } from '../../../controls'
import {
  deserializeControl,
  DeserializedControl,
  serializeControl,
  SerializedControl,
} from '../control-serialization'
import { Deserialize, Serialize } from './types'

export function serializeShapeControlDefinition(
  definition: ShapeControlDefinition,
): [Serialize<ShapeControlDefinition>, Transferable[]] {
  const [type, transferables] = Object.entries(definition.config.type).reduce(
    ([type, transferables], [key, value]) => {
      const [serializedType, serializedTransferables] = serializeControl(value)

      type[key] = serializedType
      transferables.push(...serializedTransferables)

      return [type, transferables]
    },
    [{} as Record<string, SerializedControl>, [] as Transferable[]],
  )

  return [{ ...definition, config: { ...definition.config, type } }, transferables] as [
    Serialize<ShapeControlDefinition>,
    Transferable[],
  ]
}

export function deserializeShapeControlDefinition(
  definition: Serialize<ShapeControlDefinition>,
): Deserialize<Serialize<ShapeControlDefinition>> {
  const type = Object.entries(definition.config.type).reduce((type, [key, value]) => {
    const serializedType = deserializeControl(value)

    type[key] = serializedType

    return type
  }, {} as Record<string, DeserializedControl>)

  return {
    ...definition,
    config: { ...definition.config, type },
  } as Deserialize<Serialize<ShapeControlDefinition>>
}
