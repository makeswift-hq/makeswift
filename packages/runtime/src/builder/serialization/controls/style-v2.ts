import { StyleV2ControlDefinition } from '../../../controls'
import { deserializeControl, serializeControl } from '../control-serialization'
import { deserializeFunction, serializeFunction } from '../function-serialization'
import { Serialize, Deserialize } from './types'

export function serializeStyleV2Control(
  definition: StyleV2ControlDefinition,
): [Serialize<StyleV2ControlDefinition>, Transferable[]] {
  const [type, transferables] = serializeControl(definition.config.type)
  const getStyle = definition.config.getStyle && serializeFunction(definition.config.getStyle)

  if (getStyle) {
    transferables.push(getStyle)
  }

  return [
    {
      ...definition,
      config: {
        ...definition.config,
        type,
        getStyle,
      },
    },
    transferables,
  ] as [Serialize<StyleV2ControlDefinition>, Transferable[]]
}

export function deserializeStyleV2Control(
  definition: Serialize<StyleV2ControlDefinition>,
): Deserialize<Serialize<StyleV2ControlDefinition>> {
  const type = deserializeControl(definition.config.type)
  const getStyle = definition.config.getStyle && deserializeFunction(definition.config.getStyle)

  return {
    ...definition,
    config: {
      ...definition.config,
      getStyle,
      type,
    },
  } as Deserialize<Serialize<StyleV2ControlDefinition>>
}
