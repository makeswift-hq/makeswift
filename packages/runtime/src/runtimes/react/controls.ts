import { useMemo, useRef } from 'react'

import { useStore } from '.'
import * as ReactPage from '../../state/react-page'
import { Props } from '../../prop-controllers'
import {
  Descriptor,
  Device,
  ResolveOptions,
  ResponsiveValue,
  WidthControlValueFormats,
  WidthDescriptor,
  WidthValue,
} from '../../prop-controllers/descriptors'
import { css } from '@emotion/css'
import { useResponsiveColor } from '../../components'
import type { ColorValue } from '../../components/utils/types'
import { responsiveWidth } from '../../components/utils/responsive-style'
import {
  ColorControlType,
  NumberControlType,
  SelectControlType,
  StyleControlType,
  TextAreaControlType,
  TextInputControlType,
} from '../../controls'
import { useFormattedStyle } from './controls/style'
import { useNumber } from './controls/number'
import { useTextInputValue } from './controls/text-input'
import { useTextAreaValue } from './controls/text-area'
import { useColorValue } from './controls/color'
import { useSelectControlValue } from './controls/select'

export type ResponsiveColor = ResponsiveValue<ColorValue>

function useDeviceMode(): Device {
  return 'desktop'
}

function useWidth(
  value: WidthValue | undefined,
  descriptor: WidthDescriptor,
  props: Record<string, unknown>,
): string | WidthValue | undefined {
  const deviceMode = useDeviceMode()
  const options = useMemo(
    () =>
      typeof descriptor.options === 'function'
        ? descriptor.options(props, deviceMode)
        : descriptor.options,
    [props, deviceMode],
  )

  return useMemo(
    () =>
      options.format === WidthControlValueFormats.ClassName
        ? css(responsiveWidth(value, options.defaultValue))
        : value,
    [value, options.defaultValue, options.format],
  )
}

export function useProps(element: ReactPage.ElementData): Record<string, unknown> {
  const store = useStore()
  const propControllerDescriptorsRef = useRef(
    ReactPage.getComponentPropControllerDescriptors(store.getState(), element.type) ?? {},
  )
  const props = element.props as Record<string, any>

  return Object.fromEntries(
    Object.entries(propControllerDescriptorsRef.current).map(([propName, descriptor]) => {
      switch (descriptor.type) {
        case NumberControlType:
          return [propName, useNumber(props[propName], descriptor)]

        case TextInputControlType:
          return [propName, useTextInputValue(props[propName], descriptor)]

        case TextAreaControlType:
          return [propName, useTextAreaValue(props[propName], descriptor)]

        case SelectControlType:
          return [propName, useSelectControlValue(props[propName], descriptor)]

        case ColorControlType:
          return [propName, useColorValue(props[propName], descriptor)]

        case Props.Types.ResponsiveColor: {
          const color = useResponsiveColor(props[propName])

          return [propName, color]
        }

        case Props.Types.Width:
          return [propName, useWidth(props[propName], descriptor, props)]

        case StyleControlType:
          return [propName, useFormattedStyle(props[propName], descriptor)]

        default:
          return [propName, props[propName]]
      }
    }),
  )
}

export type ResolveWidthControlValue<T extends Descriptor> = T extends WidthDescriptor
  ? undefined extends ResolveOptions<T['options']>['format']
    ? WidthValue | undefined
    : ResolveOptions<T['options']>['format'] extends typeof WidthControlValueFormats.ClassName
    ? string
    : ResolveOptions<T['options']>['format'] extends typeof WidthControlValueFormats.ResponsiveValue
    ? WidthValue | undefined
    : never
  : never
