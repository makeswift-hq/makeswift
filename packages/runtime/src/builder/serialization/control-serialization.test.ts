import { serializeControls, deserializeControls } from './control-serialization'
import { Checkbox, Number, Select } from '@makeswift/controls'

describe('deserializeControls', () => {
  test('deserializes record of serialized controls', () => {
    // Arrange
    const controls = {
      checkbox: Checkbox({ label: 'Checkbox', defaultValue: true }),
      number: Number({ label: 'Number', defaultValue: 42 }),
      select: Select({ label: 'Select', options: [{ value: 'red', label: 'Red' }] }),
    } as const

    const [serialized, transferables] = serializeControls(controls)

    // Act
    const deserializedControls = deserializeControls(serialized)

    // Assert
    expect(Object.keys(deserializedControls)).toEqual(Object.keys(controls))
    expect(deserializedControls).toEqual(controls)

    transferables.forEach((port: any) => port.close())
  })

  test('gracefully deserializes record of serialized controls when invalid config is provided', () => {
    // Arrange
    const controls = {
      checkbox: Checkbox({ label: 'Checkbox', defaultValue: true }),
      // @ts-expect-error Invalid config for number
      faultyNumber: Number({ label: 'Number', defaultValue: 'not a number!' }),
      // @ts-expect-error Invalid config for select
      faultySelect: Select({ label: 'Select', options: [{ value: 1, label: 'Red' }] }),
    } as const

    const [serialized, transferables] = serializeControls(controls)

    // Act
    const errorCallback = jest.fn()
    const deserializedControls = deserializeControls(serialized, {
      onError: errorCallback,
    })

    // Assert
    expect(Object.keys(deserializedControls)).toEqual(['checkbox'])
    expect(errorCallback).toHaveBeenCalledTimes(2)

    transferables.forEach((port: any) => port.close())
  })
})
