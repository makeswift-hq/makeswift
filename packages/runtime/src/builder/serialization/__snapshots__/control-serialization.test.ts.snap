// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deserializeControls gracefully deserializes record of serialized controls when invalid config is provided 1`] = `
[
  [Error: Could not deserialize control for "faultyCheckbox": [
  {
    "code": "invalid_type",
    "expected": "boolean",
    "received": "number",
    "path": [
      "config",
      "defaultValue"
    ],
    "message": "Expected boolean, received number"
  }
]],
  [Error: Could not deserialize control for "faultyNumber": [
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "string",
    "path": [
      "config",
      "defaultValue"
    ],
    "message": "Expected number, received string"
  }
]],
]
`;
